microcosm/loaders.py

```diff
@@ -97,30 +132,13 @@ def _load_from_environ(metadata, value_func=None):
 
     prefix = metadata.name.upper().replace("-", "_").split("_")
 
-    def matches_app(key_parts):
-        return len(key_parts) > len(prefix) and key_parts[:len(prefix)] == prefix
-
-    def assign_value(dct, key, value):
-        dct[key.lower()] = value_func(value) if value_func else value
-
-    def process_env_var(key_parts, value, config):
-        if not matches_app(key_parts):
-            return
-        dct = config
-        # walk the path specified by the env var to put the value in the right
-        # place in the config
-        for key_part in key_parts[len(prefix):-1]:
-            if key_part.lower() not in dct:
-                dct[key_part.lower()] = dict()
-            dct = dct[key_part.lower()]
-        assign_value(dct, key_parts[-1], value)
-
-    config = Configuration()
-    for key, value in environ.items():
-        separator = "__" if "__" in key else "_"
-        key_parts = key.split(separator)
-        process_env_var(key_parts, value, config)
-    return config
+    return expand_config(
+        environ,
+        separator=lambda key: "__" if "__" in key else "_",
+        skip_to=len(prefix),
+        key_parts_filter=lambda key_parts: len(key_parts) > len(prefix) and key_parts[:len(prefix)] == prefix,
```

Comment:
This logic is old, just got moved around.

The intention here is to only match environment variables that start with the right prefix so:
 - Don't match any keys that aren't long enough
 - Don't match any keys that don't have their first N parts equivalent to the prefix